# Day 2 — Pandas Deep Dive: Titanic Dataset Analysis
**Date:** 21 March 2026 | **Phase 1 — Week 1** | **Roadmap: DS-AI-75D**

## 1. Environment Setup & Data Loading

In [3]:
import pandas as pd
import numpy as np

print(f"Pandas version: {pd.__version__}")
print(f"NumPy version:  {np.__version__}")
print("Environment ready!")

Pandas version: 3.0.1
NumPy version:  2.4.3
Environment ready!


###  Load the Titanic Dataset

In [4]:
# Load Titanic dataset directly from URL
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"
df = pd.read_csv(url)

# First inspection — always do this with ANY new dataset
print("Shape:", df.shape)
print("\nColumn names:")
print(df.columns.tolist())
print("\nFirst 5 rows:")
df.head()

Shape: (891, 12)

Column names:
['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked']

First 5 rows:


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


### Dataset Inspection
Understanding our data before touching it.

In [ ]:
# Deep inspection — run this on EVERY new dataset
print("=== DATA TYPES ===")
print(df.dtypes)

print("\n=== MISSING VALUES ===")
missing = df.isnull().sum()
missing_pct = df.isnull().sum() / len(df) * 100
missing_df = pd.DataFrame({"Missing Count": missing, "Missing %": missing_pct.round(2)})
print(missing_df[missing_df["Missing Count"] > 0])

print("\n=== STATISTICAL SUMMARY ===")
df.describe()

=== DATA TYPES ===
PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

=== MISSING VALUES ===
          Missing Count  Missing %
Age                 177      19.87
Cabin               687      77.10
Embarked              2       0.22

=== STATISTICAL SUMMARY ===


,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,891.000000,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,446.000000,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,257.353842,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,223.500000,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,446.000000,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,668.500000,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


## 2. Indexing & Filtering
Three ways to select data in Pandas — master all three.

### 2.1 Column Selection

In [ ]:
# Single column → returns a Series
age = df["Age"]
print("Type:", type(age))
print("Shape:", age.shape)

# Multiple columns → returns a DataFrame
subset = df[["Name", "Age", "Sex", "Survived"]]
print("\nType:", type(subset))
print("Shape:", subset.shape)
subset.head()

Type: <class 'pandas.Series'>
Shape: (891,)

Type: <class 'pandas.DataFrame'>
Shape: (891, 4)


,Name,Age,Sex,Survived
0,"Braund, Mr. Owen Harris",22.0,male,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
2,"Heikkinen, Miss. Laina",26.0,female,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
4,"Allen, Mr. William Henry",35.0,male,0


### 2.2 loc — Label Based Selection

In [ ]:
# loc uses ROW LABELS and COLUMN NAMES
# Format: df.loc[row, column]

# Single row
print("=== Single row ===")
print(df.loc[0])

# Specific value
print("\n=== Single value ===")
print(df.loc[0, "Age"])

# Range of rows + specific columns
print("\n=== Rows 0-4, selected columns ===")
df.loc[0:4, ["Name", "Age", "Sex", "Survived"]]

=== Single row ===
PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

=== Single value ===
22.0

=== Rows 0-4, selected columns ===


,Name,Age,Sex,Survived
0,"Braund, Mr. Owen Harris",22.0,male,0
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",38.0,female,1
2,"Heikkinen, Miss. Laina",26.0,female,1
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",35.0,female,1
4,"Allen, Mr. William Henry",35.0,male,0


### 2.3 iloc — Position Based Selection

In [8]:
# iloc uses ROW NUMBERS and COLUMN NUMBERS
# Format: df.iloc[row_position, col_position]

# First row by position
print("=== First row ===")
print(df.iloc[0])

# Specific position
print("\n=== Row 0, Column 1 (Survived) ===")
print(df.iloc[0, 1])

# Range — remember EXCLUSIVE like NumPy
print("\n=== First 5 rows, first 4 columns ===")
df.iloc[0:5, 0:4]

=== First row ===
PassengerId                          1
Survived                             0
Pclass                               3
Name           Braund, Mr. Owen Harris
Sex                               male
Age                               22.0
SibSp                                1
Parch                                0
Ticket                       A/5 21171
Fare                              7.25
Cabin                              NaN
Embarked                             S
Name: 0, dtype: object

=== Row 0, Column 1 (Survived) ===
0

=== First 5 rows, first 4 columns ===


,PassengerId,Survived,Pclass,Name
0,1,0,3,"Braund, Mr. Owen Harris"
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th..."
2,3,1,3,"Heikkinen, Miss. Laina"
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)"
4,5,0,3,"Allen, Mr. William Henry"


### Key Difference: loc vs iloc
| | loc | iloc |
|---|---|---|
| Uses | Labels/Names | Numbers/Positions |
| End of range | INCLUSIVE | EXCLUSIVE |
| Example | df.loc[0:4] = 5 rows | df.iloc[0:4] = 4 rows |

### 2.4 Boolean Filtering — The Most Used Skill in Real DS Work

In [ ]:
# Single condition
survivors = df[df["Survived"] == 1]
first_class = df[df["Pclass"] == 1]
adults = df[df["Age"] >= 18]

print("=== Survival counts ===")
print(f"Total passengers:  {len(df)}")
print(f"Survivors:         {len(survivors)}")
print(f"Non-survivors:     {len(df) - len(survivors)}")
print(f"Survival rate:     {len(survivors)/len(df)*100:.1f}%")

print("\n=== Multiple conditions ===")
# AND — both conditions must be true
first_class_survivors = df[(df["Pclass"] == 1) & (df["Survived"] == 1)]
print(f"First class survivors: {len(first_class_survivors)}")

# OR — either condition can be true
children_or_elderly = df[(df["Age"] < 12) | (df["Age"] > 60)]
print(f"Children or elderly:   {len(children_or_elderly)}")

# NOT
third_class = df[~(df["Pclass"] == 3)]
print(f"Not third class:       {len(third_class)}")

=== Survival counts ===
Total passengers:  891
Survivors:         342
Non-survivors:     549
Survival rate:     38.4%

=== Multiple conditions ===
First class survivors: 136
Children or elderly:   90
Not third class:       400


## 3. GroupBy Analysis — Who Survived and Why?
Split → Apply → Combine pattern.

### Question 1: What was the overall survival rate?

In [ ]:
overall = df["Survived"].mean()
total = len(df)
survived = df["Survived"].sum()

print(f"Total passengers:  {total}")
print(f"Total survivors:   {int(survived)}")
print(f"Overall survival rate: {overall:.1%}")

Total passengers:  891
Total survivors:   342
Overall survival rate: 38.4%


### Question 2: Did passenger class determine survival?

In [ ]:
class_survival = df.groupby("Pclass")["Survived"].mean()
class_count = df.groupby("Pclass")["Survived"].count()

result = pd.DataFrame(
    {
        "Survival Rate": class_survival.map("{:.1%}".format),
        "Total Passengers": class_count,
    }
)
print(result)

       Survival Rate  Total Passengers
Pclass                                
1              63.0%               216
2              47.3%               184
3              24.2%               491


### Question 3: Did gender determine survival?

In [ ]:
gender_survival = df.groupby("Sex")["Survived"].mean()
gender_count = df.groupby("Sex")["Survived"].count()

result = pd.DataFrame(
    {
        "Survival Rate": gender_survival.map("{:.1%}".format),
        "Total Passengers": gender_count,
    }
)
print(result)

       Survival Rate  Total Passengers
Sex                                   
female         74.2%               314
male           18.9%               577


### Question 4: What was the age profile of survivors vs non-survivors?

In [ ]:
age_profile = df.groupby("Survived")["Age"].agg(["mean", "median", "std"])
age_profile.index = ["Did not survive", "Survived"]
print(age_profile.round(2))

                  mean  median    std
Did not survive  30.63    28.0  14.17
Survived         28.34    28.0  14.95


### Question 5: Which class + gender combination had best/worst survival?

In [ ]:
pivot = df.groupby(["Pclass", "Sex"])["Survived"].mean().unstack()
print("Survival rate by class and gender:")
print(pivot.map("{:.1%}".format))

Survival rate by class and gender:
Sex    female   male
Pclass              
1       96.8%  36.9%
2       92.1%  15.7%
3       50.0%  13.5%


## 4. DataFrame Merging

In [ ]:
# Create class statistics to merge back
class_stats = (
    df.groupby("Pclass")
    .agg(
        avg_survival=("Survived", "mean"),
        avg_fare=("Fare", "mean"),
        passenger_count=("PassengerId", "count"),
    )
    .reset_index()
)

print("Class statistics table:")
print(class_stats.round(2))

Class statistics table:
   Pclass  avg_survival  avg_fare  passenger_count
0       1          0.63     84.15              216
1       2          0.47     20.66              184
2       3          0.24     13.68              491


In [ ]:
# Merge class stats back into main DataFrame
df_enriched = pd.merge(df, class_stats, on="Pclass", how="left")

print("Shape before merge:", df.shape)
print("Shape after merge: ", df_enriched.shape)
print("\nSample — each passenger now has their class statistics:")
df_enriched[["Name", "Pclass", "Survived", "avg_survival", "avg_fare"]].head(8)

Shape before merge: (891, 12)
Shape after merge:  (891, 15)

Sample — each passenger now has their class statistics:


,Name,Pclass,Survived,avg_survival,avg_fare
0,"Braund, Mr. Owen Harris",3,0,0.242363,13.675550
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,1,0.629630,84.154687
2,"Heikkinen, Miss. Laina",3,1,0.242363,13.675550
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,1,0.629630,84.154687
4,"Allen, Mr. William Henry",3,0,0.242363,13.675550
5,"Moran, Mr. James",3,0,0.242363,13.675550
6,"McCarthy, Mr. Timothy J",1,0,0.629630,84.154687
7,"Palsson, Master. Gosta Leonard",3,0,0.242363,13.675550


### The Four Join Types

In [ ]:
# Create two small tables to demonstrate all join types
passengers = pd.DataFrame(
    {
        "PassengerId": [1, 2, 3, 4, 5],
        "Name": ["Allen", "Braund", "Cumings", "Heikkinen", "Futrelle"],
        "Pclass": [1, 3, 1, 3, 1],
    }
)

ticket_prices = pd.DataFrame({"Pclass": [1, 2, 3], "AvgFare": [84.15, 20.66, 13.68]})

# INNER — only matching rows from both
inner = pd.merge(passengers, ticket_prices, on="Pclass", how="inner")
print("=== INNER JOIN ===")
print(inner)

# LEFT — all from left, matching from right
left = pd.merge(passengers, ticket_prices, on="Pclass", how="left")
print("\n=== LEFT JOIN ===")
print(left)

=== INNER JOIN ===
   PassengerId       Name  Pclass  AvgFare
0            1      Allen       1    84.15
1            2     Braund       3    13.68
2            3    Cumings       1    84.15
3            4  Heikkinen       3    13.68
4            5   Futrelle       1    84.15

=== LEFT JOIN ===
   PassengerId       Name  Pclass  AvgFare
0            1      Allen       1    84.15
1            2     Braund       3    13.68
2            3    Cumings       1    84.15
3            4  Heikkinen       3    13.68
4            5   Futrelle       1    84.15


In [ ]:
# RIGHT and OUTER joins — shows NaN when no match
right = pd.merge(passengers, ticket_prices, on="Pclass", how="right")
print("=== RIGHT JOIN ===")
print(right)
# Notice: Pclass 2 appears with NaN for passenger details
# because no passenger in our small table has Pclass 2

outer = pd.merge(passengers, ticket_prices, on="Pclass", how="outer")
print("\n=== OUTER JOIN ===")
print(outer)

=== RIGHT JOIN ===
   PassengerId       Name  Pclass  AvgFare
0          1.0      Allen       1    84.15
1          3.0    Cumings       1    84.15
2          5.0   Futrelle       1    84.15
3          NaN        NaN       2    20.66
4          2.0     Braund       3    13.68
5          4.0  Heikkinen       3    13.68

=== OUTER JOIN ===
   PassengerId       Name  Pclass  AvgFare
0          1.0      Allen       1    84.15
1          3.0    Cumings       1    84.15
2          5.0   Futrelle       1    84.15
3          NaN        NaN       2    20.66
4          2.0     Braund       3    13.68
5          4.0  Heikkinen       3    13.68


## 5. Key Insights

From our Titanic analysis:

1. **Overall survival rate was 38.4%** — less than half survived
2. **Class mattered:** 1st class survived at 63% vs 24% for 3rd class
3. **Gender mattered most:** Women survived at 74% vs 19% for men
4. **Best group:** 1st class women — 96.8% survival rate
5. **Worst group:** 3rd class men — 13.5% survival rate
6. **Age:** Survivors were slightly younger (avg 28) vs non-survivors (avg 30)
7. **"Women and children first"** protocol is clearly visible in the data